include/seqan3/io/stream/parse_condition.hpp

```diff
@@ -0,0 +1,468 @@
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+
+/*!\file
+ * \brief Provides parse conditions for tokenization.
+ * \author Rene Rahn <rene.rahn AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <seqan3/alphabet/all.hpp>
+#include <seqan3/core/concept/core.hpp>
+#include <seqan3/core/detail/reflection.hpp>
+#include <seqan3/range/container/constexpr_string.hpp>
+
+#include <cctype>
+#include <string>
+
+namespace seqan3::detail
+{
+
+// ----------------------------------------------------------------------------
+// condition_message_v
+// ----------------------------------------------------------------------------
+
+/*!\brief Defines a compound seqan3::detail::constexpr_string consisting of all given conditions separated by the
+ *        operator-name `op`.
+ * \ingroup io
+ * \tparam op               non-type template parameter specifying the separator character, e.g. '|'.
+ * \tparam condition_head_t The first condition type in the message. Ensures, that there is at least one type.
+ * \tparam condition_ts     Remaining list of conditions separated by `op`.
+ */
+template <char op, typename condition_head_t, typename ...condition_ts>
+constexpr constexpr_string condition_message_v
+{
+    constexpr_string{"("} +
+    (condition_head_t::msg + ... +
+        (constexpr_string{" "} + constexpr_string{{op, op, '\0'}} + constexpr_string{" "} + condition_ts::msg)) +
+    constexpr_string{")"}
+};
+
+// ----------------------------------------------------------------------------
+// parse_condition_concept
+// ----------------------------------------------------------------------------
+
+/*!\interface seqan3::detail::parse_condition_concept <>
+ * \brief An internal concept to check if an object fulfills the requirements of a seqan3::detail::parse_condition.
+ * \ingroup io
+ *
+ * \details
+ *
+ * The must be invocable with an seqan3::alphabet_concept type and supply a static constexpr `msg` member of type
+ * seqan3::detail::constexpr_string.
+ */
+//!\cond
+template <typename condition_t>
+concept bool parse_condition_concept = invocable_concept<std::remove_reference_t<condition_t>, char> &&
+                                       requires { std::remove_reference_t<condition_t>::msg; };
+//!\endcond
+
+// ----------------------------------------------------------------------------
+// parse_condition
+// ----------------------------------------------------------------------------
+
+//!\cond
+template <parse_condition_concept... condition_ts>
+    requires sizeof...(condition_ts) >= 2
+struct or_fn;
+
+template <parse_condition_concept... condition_ts>
+    requires sizeof...(condition_ts) >= 2
+struct and_fn;
+
+template <parse_condition_concept condition_t>
+struct not_fn;
+//!\endcond
+
+/*!\brief A CRTP class for parse conditions to add logical conjunction, disjunction and negation operator.
```

Comment:
please explain `CRTP` or at least write it in full words

Reply:
or link an article about this, e.g. https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern